### ISSUE TITLE
HELP: Override root element name

### ISSUE THREAD

**Issue by Marvin-Brouwer** [YAXLib/YAXLib#163]
**Is your feature request related to a problem? Please describe.**
Hi, I have a scenario where I need to serialize different classes to similar XML like so:

```cs
public class Project {

        [YAXSerializeAs("id")]
        public string Id{ get; set; }

        [YAXSerializeAs("name")]
        public string Name { get; set; }
}

public class CreateProjectRequest : CreateRequest<Project> {

        [YAXSerializeAs("name")]
        public string Name { get; set; }
}

public class UpdateProjectRequest : UpdateRequest<Project> {

        [YAXSerializeAs("id")]
        public string Id{ get; set; }

        [YAXSerializeAs("name")]
        public string Name { get; set; }
}
```
All of the above are supposed to correspond to the followin xml:

```xml
<request>
  <Read>
    <Project>
       <id>SomeId</id>
    </Project>
  </Read>
<request>

<request>
  <Create>
    <Project>
       <name>SomeName</name>
    </Project>
  </Create>
<request>

<request>
  <Update>
    <Project>
       <id>SomeId</id>
       <name>SomeName update</name>
    </Project>
  </Update>
<request>
```

Because of the nature of the API our Update class has a `List<UpdateRequest<TModel>>` for example.
Of this, the concrete implementation may be the `UpdateProjectRequest` but because of the type constraint, YAXLib converts this to `UpdateRequestOfProject` which is to be expected.
However, I'd like to override this by some means.

I tried the `ICustomSerializer<TModel>` and I managed to change the `ElementName` but now I'm required to manually serialize all of the properties, which are not the same for all our models.

**Describe the solution you'd like**
I would be happy with any of these options:
- Telling me that this is already possible and I misunderstood the docs
- An attribute I can put on a method to override the root element
- An option to implement an alternative to `ICustomSerializer<TModel>` with an already filled object 
- An option to implement something like CustomSerializer<TModel> with a base implementation that does everything I need to happen before I override the ElementName.
- I guess a custom known type implementation could be an option too:
  https://github.com/YAXLib/YAXLib/issues/93

**Describe alternatives you've considered**
I tried the `ICustomSerializer<TModel>` and I managed to change the `ElementName` but now I'm required to manually serialize all of the properties, which are not the same for all our models.

I also browsed through the source code and I couldn't find a good way to override the RootElementName.

**Comment by axunonb** [comment]
Hi Marvin,
Thanks for your feature request, and for even diving into the *YAXLib* source code.
In order figure out all aspects (e.g. how is `CreateRequest<Project>` or `UpdateRequest<Project>` implemented), a (small) working sample would be more than helpful, also regarding current vs. expected result.
Decorating the classes with `[YAXSerializeAs("something")]` is nothing pointing into the right direction?

**Comment by Marvin-Brouwer** [comment]
Hi @axunonb,

I'm not sure if I'm allowed to share that much detail about our project. 
However, I have found a solution to fix this, or rather hack this.
In short all I needed was a way to override the root xml element of the items in a collection, so I made this:

```cs
public abstract class UpdateRequest<TModel>
        where TModel : RedactedModel
    {

        [YAXSerializeAs("id")]
        public string Id { get; set; }

        [YAXCustomSerializer(typeof(RootSerializer))]
        public abstract string XmlRoot { get; }
    }

public sealed class RootSerializer : ICustomSerializer<string>
    {
        public void SerializeToAttribute(string objectToSerialize, XAttribute attrToFill)
        {
            throw new NotSupportedException();
        }

        public void SerializeToElement(string objectToSerialize, XElement elemToFill)
        {
            elemToFill.Parent!.Name = objectToSerialize;
            elemToFill.Remove();

        }

        public string SerializeToValue(string objectToSerialize)
        {
            throw new NotSupportedException();
        }

        public string DeserializeFromAttribute(XAttribute attrib)
        {
            throw new NotSupportedException();
        }

        public string DeserializeFromElement(XElement element)
        {
            throw new NotSupportedException();
        }

        public string DeserializeFromValue(string value)
        {
            throw new NotSupportedException();
        }
    }
```

This is not the prettiest of options but it works for now.
I'd still rather have a way of updating the root element by an attribute or something on a type itself instead of on a property of type string.

**Comment by Marvin-Brouwer** [comment]
I am having a similar issue the other way around though.
I have a very specific List I'd like control over the deserialization but I'd like the items to use the original YAXSerializer.
I guess I can create a new instance but that feels dirty.

What do you think about doing what JSON.Net does, add a reference to the serializer in the methods of a CustomSerializer?
```cs
public T DeserializeFromElement(XElement element, YAXSerializer serializer)
        {
            throw new NotSupportedException();
        }
```

That way you have everything you already configured and you can modify collections, root items etc and then foreach element.Elements() you can run serializer.Serialize(child)

**Comment by axunonb** [comment]
https://github.com/YAXLib/YAXLib/issues/93 will be resolved in v4:
`IKnowType` and `KnownTypes` will be public.

**Comment by axunonb** [comment]
Resolved with release [v4.0.0](https://github.com/YAXLib/YAXLib/releases/tag/v4.0.0)
